pyccel/stdlib/cwrapper/cwrapper.c

```diff
@@ -0,0 +1,508 @@
+ */
+
+PyObject	*Complex128_to_PyComplex(double complex *c)
+{
+	double		real_part;
+	double		imag_part;
+	PyObject	*object;
+
+	real_part = creal(*c);
+	imag_part = cimag(*c);
+	object = PyComplex_FromDoubles(real_part, imag_part);
+
+	return object;
+}
+//-----------------------------------------------------//
+PyObject	*Complex64_to_PyComplex(float complex *c)
+{
+	float		real_part;
+	float		imag_part;
+	PyObject	*object;
+
+	real_part = crealf(*c);
+	imag_part = cimagf(*c);
+	object = PyComplex_FromDoubles((double) real_part, (double) imag_part);
+
+	return object;
+}
+//-----------------------------------------------------//
+PyObject	*Bool_to_PyBool(bool *b)
+{
+	return *b == true ? Py_True : Py_False;
+}
+//-----------------------------------------------------//
+PyObject	*Int64_to_PyLong(int64_t *i)
+{
+	PyObject	*object;
+
+	object = PyLong_FromLongLong((long long) *i);
+
+	return object;
+}
+//-----------------------------------------------------//
+PyObject	*Int32_to_PyLong(int32_t *i)
+{
+	PyObject	*object;
+
+	object = PyLong_FromLongLong((long long) *i);
+
+	return object;
+}
+//-----------------------------------------------------//
+PyObject	*Int16_to_PyLong(int16_t *i)
+{
+	PyObject	*object;
+
+	object = PyLong_FromLongLong((long long) *i);
+
+	return object;
+}
+//--------------------------------------------------------//
+PyObject	*Int8_to_PyLong(int8_t *i)
+{
+	PyObject	*object;
+
+	object = PyLong_FromLongLong((long long) *i);
+
+	return object;
+}
+//--------------------------------------------------------//
+PyObject	*Double_to_PyDouble(double *d)
+{
+	PyObject	*object;
+
+	object = PyFloat_FromDouble(*d);
+
+	return object;
+}
+//--------------------------------------------------------//
+PyObject	*Float_to_PyDouble(float *d)
+{
+	PyObject	*object;
+
+	object = PyFloat_FromDouble((double)*d);
+
+	return object;
+}
+
+/*
+ * Functions : Cast functions
+ * ---------------------------
+ * Some of the function used below are based on C/python api and numpy/c api with
+ * more tolerance to different precisions, different system architectures and complex type.
+ * Check the C data type ob a python object
+ * Parameters :
+ *     object     : the python object
+ *     hard_check : boolean true if intensive precision check is needed
+ *
+ * Returns    :
+ *     boolean : logic statement responsible for checking python data type
+ * Reference of the used c/python api function
+ * ---------------------------------------------------
+ * https://docs.python.org/3/c-api/long.html#c.PyLong_Check
+ * https://docs.python.org/3/c-api/complex.html#c.PyComplex_Check
+ * https://docs.python.org/3/c-api/float.html#c.PyFloat_Check
+ * https://docs.python.org/3/c-api/bool.html#c.PyBool_Check
+ * https://numpy.org/doc/1.17/reference/c-api.array.html#c.PyArray_IsScalar
+ */
+
+bool    PyIs_Int8(PyObject *o, bool hard_check)
+{
+	if (hard_check == true)
+		return PyArray_IsScalar(o, Int8);
```

Comment:
please can you explain what does hard check stand for?

Reply:
Given the code it seems that `hard_check` indicates whether it is assumed that the type is a numpy type or whether conversions are allowed

Reply:
This code should not do `hard_check == true`. Comparison of a boolean to true is a code smell

Reply:
https://github.com/pyccel/pyccel/blob/wrapper_part_1/pyccel/stdlib/cwrapper/cwrapper.c#L250-L257

Reply:
I am also not clear on why `hard_check` is needed. I don't know when it is set to `False`. It also seems that this argument is always set literally and never parametrically, so I would rather break these up into 2 functions to avoid unnecessary switches. Maybe we can ask @MenoIy to tell us more about this parameter?

Reply:
> I would rather break these up into 2 functions to avoid unnecessary switches

Done 47b2c39

Reply:
After discussion, this parameter is not necessary. It has been removed: b8cc4c5

Reply:
You did a great job @EmilyBourne. finally, we can work on many issues blocked by the wrapper.